aries_vcx/src/handlers/connection/invitee/state_machine.rs

```diff
@@ -400,6 +399,16 @@ impl SmConnectionInvitee {
     pub fn handle_ack(self, _ack: Ack) -> VcxResult<Self> {
         Ok(self)
     }
+
+    pub fn get_thread_id(&self) -> VcxResult<String> {
+        match &self.state {
+            InviteeFullState::Invited(_) => Err(VcxError::from_msg(VcxErrorKind::NotReady, "Thread ID not yet available in this state")),
+            InviteeFullState::Requested(state) => Ok(state.request.id.0.clone()),
+            InviteeFullState::Responded(state) => state.response.thread.thid.clone().ok_or(VcxError::from_msg(VcxErrorKind::UnknownError, "Thread ID missing on connection")),
```

Comment:
I understand this error handling arises from general optionality of thid by aries spec, but I think we should not allow this in a first place. Receiving connection response without thread id should not be valid. And if we allow it, we probably only delay issues (making them harder to debug), as for example, `get_thread_id`  will be failing (because we have previously allowed `connection response` without `thid`)

Reply:
We should have way to express that in some cases, `thid` is required. Maybe have 2 `Thread` structures, where one has `thid` required and other optional? Doesn't sound great but seems to me sometimes we need to be able to express that `thid` is required.

Reply:
I know we had this discussion before and to be honest I don't remember 100% of it - but would we break something if we just make `thid` required? Do we have case is `aries-vcx` implementation where we construct thread structure without `thid`?

Reply:
I don't understand why `thid` being optional is a problem as long as, if it is expected, it is verified immediately after receiving the message, which it is.

Reply:
It allows for scenarios which should never happen - in messages and scenarios where `thid` is required (basically almost always at normal circumstances, except for initial messages or problems reports), typing `thid` as Option does not reflect the requirement and delegates the responsibility to check value of `Option` to the developer anytime you try to access this data. Just like in the case above:

```
InviteeFullState::Responded(state) => state.response.thread.thid.clone().ok_or(VcxError::from_msg(VcxErrorKind::UnknownError, "Thread ID missing on connection")),
```

we need to throw `Uknown Error` because we don't really know what happened - we as developer must have made mistake if this happens - we know we are checking `Connection Response` for `thid` anytime we use it to update state.   If `thid` in this case was not optional, either:
1. it has `thid`, we continue and we never have to worry about throwing `VcxErrorKind::UnknownError` elsewhere in code
2. it does not have `thid` and we fail quickly. 

I mean - at core making a field mandatory if it is mandatory just reflects reality better and that itself then have positive effects on code cleaniness. 

The threading RFC states that `thid` is optional but most of the time it's not. I am not sure what'd be the best technical solution - the simplest I can think of is having 2 `Thread` structures, one with `thid` optional, the other not - perhaps `GeneralThread` and `ConversationThread`. You would use `Conversation Thread` on messages which are expected to be part of a conversation (must have `thid`) and `GeneralThread` in cases when AriesRFC-wide flexibility is required.